### ISSUE TITLE
VCC performance regression in 6.3 and 6.4.0

### ISSUE THREAD

**Issue by n4n0r** [varnishcache/varnish-cache#3392]
Hi Varnish team !

- Following up a previous conversation. In the process of upgrading from varnish 6.2.2 to 6.4.0, I noticed compilation of the VCL was taking much longer with varnish 6.4.0 (7-10x longer). It turns out this is the case with 6.3 too.
In research of a possible VCC performance regression, I did run a git bisect and manage to find the responsible commit of the increase in compilation time:

```
8bf0a33aee49e10b01dd9aea54a6c01e60bc85d5 is the first bad commit
commit 8bf0a33aee49e10b01dd9aea54a6c01e60bc85d5
Author: Poul-Henning Kamp <phk@FreeBSD.org>
Date:   Tue Jun 18 07:07:47 2019 +0000

    Convert std.log() and std.syslog from STRING_LIST to STRANDS.
```
### $ git bisect log
```
git bisect start
# good: [3ed1506895ecaddb91f658bee11742f0b0b982b5] Prepare for 6.2.2
git bisect good 3ed1506895ecaddb91f658bee11742f0b0b982b5
# bad: [6e96ff048692235e64565211a38c41432a26c055] Prepare for 6.3.1
git bisect bad 6e96ff048692235e64565211a38c41432a26c055
# good: [d7e64e2f76bcc7ef6932af89c1980f0c6e9290c5] VRT is now officially at 9.0
git bisect good d7e64e2f76bcc7ef6932af89c1980f0c6e9290c5
# good: [c611b9a615b2d035c8631febc9f60efb339af67e] Guard against OA_GZIPBITS streaming race
git bisect good c611b9a615b2d035c8631febc9f60efb339af67e
# bad: [bb8e3ec75924264ee1e417540bf4f1a387c7301b] Adjust for other OS
git bisect bad bb8e3ec75924264ee1e417540bf4f1a387c7301b
# good: [c1990ad9f9bee5a8532ddd88765fb823ebf3f439] Necessary #include for previous commit.
git bisect good c1990ad9f9bee5a8532ddd88765fb823ebf3f439
# good: [7e14355f76dba7fe77043085bf7a08bbe20f8852] Convert std.toupper() and std.tolower() from STRING_LIST to STRANDS
git bisect good 7e14355f76dba7fe77043085bf7a08bbe20f8852
# bad: [5ecbde3812f9d17b47faf3b6c44567bcb7060fac] Make a special case for sending h2 RST frames
git bisect bad 5ecbde3812f9d17b47faf3b6c44567bcb7060fac
# bad: [efa582f5e8073fdc95e15d450bf7e112071febff] Add VRT_Strands2Bool() so we dont reassemble a multi-part string on the workspace to check if any element of it is non-NULL.
git bisect bad efa582f5e8073fdc95e15d450bf7e112071febff
# bad: [287213b11c8b1bc5fa1e135c536eadc8dfb909ea] Add a missing '='
git bisect bad 287213b11c8b1bc5fa1e135c536eadc8dfb909ea
# bad: [7c3fdb42b53227d816d429d9231f4c6b39d1bfed] Move VRT_BundleStrands() to VPI_BundleStrands()
git bisect bad 7c3fdb42b53227d816d429d9231f4c6b39d1bfed
# bad: [8bf0a33aee49e10b01dd9aea54a6c01e60bc85d5] Convert std.log() and std.syslog from STRING_LIST to STRANDS.
git bisect bad 8bf0a33aee49e10b01dd9aea54a6c01e60bc85d5
# first bad commit: [8bf0a33aee49e10b01dd9aea54a6c01e60bc85d5] Convert std.log() and std.syslog from STRING_LIST to STRANDS.
```

FYI, the VCL config is fairly big and contains thousands of std.log. Before the regression the VCL compiles in about 2m50, after in about 23 minutes. (no stack protector make things compile normally, but unfortunatly, we can;t deactivate this).
Compilation is done without optimisation (-o0) to reduce compilation speed, here is the compile line:
```
'cc_command=exec gcc -std=gnu99 -g -fstack-protector -fno-var-tracking -O0 --param=ssp-buffer-size=4 -Wformat -Wformat-security -Werror=format-security -Wall -Werror -Wno-error=unused-result -pthread -fpic -shared -Wl,-x -o %o %s'
```

-  Additionally, for extra information, find a ftime-report comparison of the compilation of this VCL between 6.2.2 and 6.4.0:
### VARNISH 6.2.2
```
# /usr/lib/gcc/x86_64-linux-gnu/5/cc1 -quiet -imultiarch x86_64-linux-gnu -D_REENTRANT vgc.c -quiet -dumpbase vgc.c -mtune=generic -march=x86-64 -auxbase vgc -g -O0 -Wformat=1 -Wformat-security -Werror=format-security -Wall -Werror -Wno-error=unused-result -std=gnu99 -fstack-protector -fno-var-tracking -fpic -ftime-report --param ssp-buffer-size=4 -Wformat-security
Execution times (seconds)
 phase setup             :   0.00 ( 0%) usr   0.00 ( 0%) sys   0.00 ( 0%) wall    1104 kB ( 0%) ggc
 phase parsing           :   9.88 ( 5%) usr   7.98 (67%) sys  18.02 ( 9%) wall  292534 kB (13%) ggc
 phase lang. deferred    :   0.02 ( 0%) usr   0.03 ( 0%) sys   0.04 ( 0%) wall       0 kB ( 0%) ggc
 phase opt and generate  : 170.29 (94%) usr   3.87 (32%) sys 182.30 (91%) wall 1981162 kB (86%) ggc
 phase debug info        :   0.03 ( 0%) usr   0.01 ( 0%) sys   0.04 ( 0%) wall   27367 kB ( 1%) ggc
 phase last asm          :   0.08 ( 0%) usr   0.00 ( 0%) sys   0.08 ( 0%) wall     759 kB ( 0%) ggc
 phase finalize          :   0.00 ( 0%) usr   0.00 ( 0%) sys   0.01 ( 0%) wall       0 kB ( 0%) ggc
 garbage collection      :   1.45 ( 1%) usr   0.00 ( 0%) sys   1.48 ( 1%) wall       0 kB ( 0%) ggc
 dump files              :   0.00 ( 0%) usr   0.00 ( 0%) sys   0.01 ( 0%) wall       0 kB ( 0%) ggc
 callgraph construction  :   1.22 ( 1%) usr   0.08 ( 1%) sys   1.28 ( 1%) wall  173986 kB ( 8%) ggc
 callgraph optimization  :   0.16 ( 0%) usr   0.00 ( 0%) sys   0.22 ( 0%) wall      52 kB ( 0%) ggc
 ipa dead code removal   :   0.09 ( 0%) usr   0.00 ( 0%) sys   0.08 ( 0%) wall       0 kB ( 0%) ggc
 ipa inlining heuristics :   0.10 ( 0%) usr   0.01 ( 0%) sys   0.16 ( 0%) wall       0 kB ( 0%) ggc
 ipa free inline summary :   0.01 ( 0%) usr   0.00 ( 0%) sys   0.01 ( 0%) wall       0 kB ( 0%) ggc
 cfg construction        : 129.72 (72%) usr   0.47 ( 4%) sys 132.76 (66%) wall   95640 kB ( 4%) ggc
 cfg cleanup             :   1.05 ( 1%) usr   0.04 ( 0%) sys   1.15 ( 1%) wall    5052 kB ( 0%) ggc
 trivially dead code     :   0.55 ( 0%) usr   0.00 ( 0%) sys   0.57 ( 0%) wall       0 kB ( 0%) ggc
 df scan insns           :   2.35 ( 1%) usr   0.36 ( 3%) sys   2.85 ( 1%) wall       3 kB ( 0%) ggc
 df live regs            :   1.40 ( 1%) usr   0.05 ( 0%) sys   1.48 ( 1%) wall       0 kB ( 0%) ggc
 df reg dead/unused notes:   1.23 ( 1%) usr   0.00 ( 0%) sys   1.28 ( 1%) wall   47358 kB ( 2%) ggc
 register information    :   0.72 ( 0%) usr   0.03 ( 0%) sys   0.74 ( 0%) wall       0 kB ( 0%) ggc
 alias analysis          :   0.32 ( 0%) usr   0.09 ( 1%) sys   0.46 ( 0%) wall   15344 kB ( 1%) ggc
 rebuild jump labels     :   0.28 ( 0%) usr   0.00 ( 0%) sys   0.33 ( 0%) wall       0 kB ( 0%) ggc
 preprocessing           :   2.75 ( 2%) usr   2.07 (17%) sys   4.93 ( 2%) wall  105697 kB ( 5%) ggc
 lexical analysis        :   4.35 ( 2%) usr   3.42 (29%) sys   7.52 ( 4%) wall       0 kB ( 0%) ggc
 parser (global)         :   0.62 ( 0%) usr   0.50 ( 4%) sys   1.41 ( 1%) wall   51283 kB ( 2%) ggc
 parser function body    :   2.01 ( 1%) usr   1.99 (17%) sys   4.00 ( 2%) wall  134744 kB ( 6%) ggc
 early inlining heuristics:   0.00 ( 0%) usr   0.00 ( 0%) sys   0.03 ( 0%) wall       0 kB ( 0%) ggc
 inline parameters       :   0.66 ( 0%) usr   0.00 ( 0%) sys   0.63 ( 0%) wall   19269 kB ( 1%) ggc
 tree gimplify           :   1.15 ( 1%) usr   0.04 ( 0%) sys   1.21 ( 1%) wall  368646 kB (16%) ggc
 tree eh                 :   0.21 ( 0%) usr   0.00 ( 0%) sys   0.20 ( 0%) wall       3 kB ( 0%) ggc
 tree CFG construction   :   0.48 ( 0%) usr   0.02 ( 0%) sys   0.48 ( 0%) wall  128766 kB ( 6%) ggc
 tree CFG cleanup        :   0.39 ( 0%) usr   0.01 ( 0%) sys   0.42 ( 0%) wall       4 kB ( 0%) ggc
 tree PHI insertion      :   0.32 ( 0%) usr   0.01 ( 0%) sys   0.32 ( 0%) wall   31482 kB ( 1%) ggc
 tree SSA rewrite        :   0.29 ( 0%) usr   0.01 ( 0%) sys   0.30 ( 0%) wall   79995 kB ( 3%) ggc
 tree SSA other          :   0.68 ( 0%) usr   0.43 ( 4%) sys   1.04 ( 1%) wall       5 kB ( 0%) ggc
 tree operand scan       :   0.54 ( 0%) usr   0.45 ( 4%) sys   1.07 ( 1%) wall   66220 kB ( 3%) ggc
 dominance frontiers     :   0.06 ( 0%) usr   0.00 ( 0%) sys   0.02 ( 0%) wall       0 kB ( 0%) ggc
 dominance computation   :   0.76 ( 0%) usr   0.03 ( 0%) sys   0.76 ( 0%) wall       0 kB ( 0%) ggc
 out of ssa              :   0.37 ( 0%) usr   0.00 ( 0%) sys   0.39 ( 0%) wall    2495 kB ( 0%) ggc
 expand vars             :   0.12 ( 0%) usr   0.00 ( 0%) sys   0.09 ( 0%) wall   28767 kB ( 1%) ggc
 expand                  :   1.69 ( 1%) usr   0.22 ( 2%) sys   2.08 ( 1%) wall  528147 kB (23%) ggc
 post expand cleanups    :   0.33 ( 0%) usr   0.00 ( 0%) sys   0.35 ( 0%) wall      16 kB ( 0%) ggc
 varconst                :   0.02 ( 0%) usr   0.00 ( 0%) sys   0.04 ( 0%) wall     337 kB ( 0%) ggc
 loop init               :   0.15 ( 0%) usr   0.00 ( 0%) sys   0.15 ( 0%) wall      11 kB ( 0%) ggc
 loop fini               :   0.01 ( 0%) usr   0.00 ( 0%) sys   0.00 ( 0%) wall       0 kB ( 0%) ggc
 integrated RA           :   9.99 ( 6%) usr   0.93 ( 8%) sys  12.83 ( 6%) wall  264581 kB (11%) ggc
 LRA non-specific        :   3.67 ( 2%) usr   0.14 ( 1%) sys   4.93 ( 2%) wall    1073 kB ( 0%) ggc
 LRA virtuals elimination:   0.80 ( 0%) usr   0.01 ( 0%) sys   1.02 ( 1%) wall   37777 kB ( 2%) ggc
 LRA reload inheritance  :   0.02 ( 0%) usr   0.00 ( 0%) sys   0.01 ( 0%) wall       0 kB ( 0%) ggc
 LRA create live ranges  :   0.86 ( 0%) usr   0.17 ( 1%) sys   1.73 ( 1%) wall    4930 kB ( 0%) ggc
 LRA hard reg assignment :   0.06 ( 0%) usr   0.05 ( 0%) sys   0.50 ( 0%) wall       0 kB ( 0%) ggc
 reload                  :   0.08 ( 0%) usr   0.01 ( 0%) sys   0.19 ( 0%) wall       0 kB ( 0%) ggc
 thread pro- & epilogue  :   1.11 ( 1%) usr   0.03 ( 0%) sys   1.16 ( 1%) wall      69 kB ( 0%) ggc
 machine dep reorg       :   0.06 ( 0%) usr   0.00 ( 0%) sys   0.07 ( 0%) wall       0 kB ( 0%) ggc
 shorten branches        :   1.00 ( 1%) usr   0.01 ( 0%) sys   1.06 ( 1%) wall       0 kB ( 0%) ggc
 final                   :   2.38 ( 1%) usr   0.16 ( 1%) sys   2.73 ( 1%) wall   79773 kB ( 3%) ggc
 symout                  :   0.11 ( 0%) usr   0.01 ( 0%) sys   0.16 ( 0%) wall   28603 kB ( 1%) ggc
 uninit var analysis     :   0.04 ( 0%) usr   0.00 ( 0%) sys   0.06 ( 0%) wall       0 kB ( 0%) ggc
 rest of compilation     :   1.47 ( 1%) usr   0.01 ( 0%) sys   1.68 ( 1%) wall    1609 kB ( 0%) ggc
 unaccounted todo        :   0.01 ( 0%) usr   0.00 ( 0%) sys   0.01 ( 0%) wall       0 kB ( 0%) ggc
 repair loop structures  :   0.03 ( 0%) usr   0.00 ( 0%) sys   0.02 ( 0%) wall       0 kB ( 0%) ggc
 TOTAL                 : 180.30            11.91           200.51            2302944 kB
```
### VARNISH 6.4.0:
```
# /usr/lib/gcc/x86_64-linux-gnu/5/cc1 -quiet -imultiarch x86_64-linux-gnu -D_REENTRANT vgc.c -quiet -dumpbase vgc.c -mtune=generic -march=x86-64 -auxbase vgc -g -O0 -Wformat=1 -Wformat-security -Werror=format-security -Wall -Werror -Wno-error=unused-result -std=gnu99 -fstack-protector -fno-var-tracking -fpic -ftime-report --param ssp-buffer-size=4 -Wformat-security -o /tmp/test.s    
Execution times (seconds)
 phase setup             :   0.00 ( 0%) usr   0.00 ( 0%) sys   0.00 ( 0%) wall    1104 kB ( 0%) ggc
 phase parsing           :   9.95 ( 1%) usr   9.15 (64%) sys  19.11 ( 1%) wall  319879 kB (12%) ggc
 phase lang. deferred    :   0.03 ( 0%) usr   0.01 ( 0%) sys   0.03 ( 0%) wall       0 kB ( 0%) ggc
 phase opt and generate  :1410.95 (99%) usr   5.11 (36%) sys1445.52 (99%) wall 2286345 kB (87%) ggc
 phase debug info        :   0.06 ( 0%) usr   0.02 ( 0%) sys   0.18 ( 0%) wall   28424 kB ( 1%) ggc
 phase last asm          :   0.12 ( 0%) usr   0.00 ( 0%) sys   0.17 ( 0%) wall    1322 kB ( 0%) ggc
 phase finalize          :   0.00 ( 0%) usr   0.00 ( 0%) sys   0.05 ( 0%) wall       0 kB ( 0%) ggc
 garbage collection      :   1.47 ( 0%) usr   0.00 ( 0%) sys   1.49 ( 0%) wall       0 kB ( 0%) ggc
 callgraph construction  :   1.28 ( 0%) usr   0.04 ( 0%) sys   1.32 ( 0%) wall  214673 kB ( 8%) ggc
 callgraph optimization  :   0.19 ( 0%) usr   0.01 ( 0%) sys   0.19 ( 0%) wall      52 kB ( 0%) ggc
 ipa dead code removal   :   0.08 ( 0%) usr   0.00 ( 0%) sys   0.08 ( 0%) wall       0 kB ( 0%) ggc
 ipa inlining heuristics :   0.22 ( 0%) usr   0.28 ( 2%) sys   4.11 ( 0%) wall       0 kB ( 0%) ggc
 ipa free inline summary :   0.01 ( 0%) usr   0.00 ( 0%) sys   0.01 ( 0%) wall       0 kB ( 0%) ggc
 cfg construction        : 107.24 ( 8%) usr   0.21 ( 1%) sys 108.22 ( 7%) wall   99697 kB ( 4%) ggc
 cfg cleanup             :   1.09 ( 0%) usr   0.01 ( 0%) sys   1.28 ( 0%) wall    5052 kB ( 0%) ggc
 trivially dead code     :   0.56 ( 0%) usr   0.03 ( 0%) sys   0.62 ( 0%) wall       0 kB ( 0%) ggc
 df scan insns           :   2.48 ( 0%) usr   0.51 ( 4%) sys   3.89 ( 0%) wall       3 kB ( 0%) ggc
 df multiple defs        :   0.01 ( 0%) usr   0.00 ( 0%) sys   0.00 ( 0%) wall       0 kB ( 0%) ggc
 df live regs            :   1.37 ( 0%) usr   0.06 ( 0%) sys   1.53 ( 0%) wall       0 kB ( 0%) ggc
 df live&initialized regs:   0.00 ( 0%) usr   0.00 ( 0%) sys   0.02 ( 0%) wall       0 kB ( 0%) ggc
 df reg dead/unused notes:   1.23 ( 0%) usr   0.02 ( 0%) sys   1.30 ( 0%) wall   52725 kB ( 2%) ggc
 register information    :   0.69 ( 0%) usr   0.09 ( 1%) sys   0.92 ( 0%) wall       0 kB ( 0%) ggc
 alias analysis          :   0.35 ( 0%) usr   0.02 ( 0%) sys   0.42 ( 0%) wall   16593 kB ( 1%) ggc
 alias stmt walking      :   0.02 ( 0%) usr   0.00 ( 0%) sys   0.03 ( 0%) wall       0 kB ( 0%) ggc
 rebuild jump labels     :   0.32 ( 0%) usr   0.01 ( 0%) sys   0.35 ( 0%) wall       0 kB ( 0%) ggc
 preprocessing           :   2.46 ( 0%) usr   2.14 (15%) sys   4.98 ( 0%) wall  108057 kB ( 4%) ggc
 lexical analysis        :   4.17 ( 0%) usr   4.14 (29%) sys   8.44 ( 1%) wall       0 kB ( 0%) ggc
 parser (global)         :   0.80 ( 0%) usr   0.53 ( 4%) sys   1.07 ( 0%) wall   51793 kB ( 2%) ggc
 parser function body    :   2.37 ( 0%) usr   2.32 (16%) sys   4.43 ( 0%) wall  159209 kB ( 6%) ggc
 early inlining heuristics:   0.01 ( 0%) usr   0.00 ( 0%) sys   0.01 ( 0%) wall       0 kB ( 0%) ggc
 inline parameters       :   0.67 ( 0%) usr   0.01 ( 0%) sys   0.74 ( 0%) wall   19269 kB ( 1%) ggc
 tree gimplify           :   1.29 ( 0%) usr   0.02 ( 0%) sys   1.30 ( 0%) wall  416919 kB (16%) ggc
 tree eh                 :   0.35 ( 0%) usr   0.00 ( 0%) sys   0.35 ( 0%) wall   28601 kB ( 1%) ggc
 tree CFG construction   :   0.50 ( 0%) usr   0.04 ( 0%) sys   0.53 ( 0%) wall  146529 kB ( 6%) ggc
 tree CFG cleanup        :   0.41 ( 0%) usr   0.03 ( 0%) sys   0.59 ( 0%) wall       4 kB ( 0%) ggc
 tree PTA                :   0.01 ( 0%) usr   0.00 ( 0%) sys   0.02 ( 0%) wall     152 kB ( 0%) ggc
 tree PHI insertion      :   0.34 ( 0%) usr   0.04 ( 0%) sys   0.35 ( 0%) wall   35518 kB ( 1%) ggc
 tree SSA rewrite        :   0.35 ( 0%) usr   0.01 ( 0%) sys   0.36 ( 0%) wall   94051 kB ( 4%) ggc
 tree SSA other          :   0.74 ( 0%) usr   0.46 ( 3%) sys   1.12 ( 0%) wall       5 kB ( 0%) ggc
 tree operand scan       :   0.64 ( 0%) usr   0.36 ( 3%) sys   1.10 ( 0%) wall   74556 kB ( 3%) ggc
 dominator optimization  :   0.02 ( 0%) usr   0.00 ( 0%) sys   0.02 ( 0%) wall     145 kB ( 0%) ggc
 tree CCP                :   0.01 ( 0%) usr   0.00 ( 0%) sys   0.02 ( 0%) wall       0 kB ( 0%) ggc
 tree FRE                :   0.01 ( 0%) usr   0.00 ( 0%) sys   0.03 ( 0%) wall     243 kB ( 0%) ggc
 tree linearize phis     :   0.00 ( 0%) usr   0.01 ( 0%) sys   0.00 ( 0%) wall       2 kB ( 0%) ggc
 tree conservative DCE   :   0.01 ( 0%) usr   0.00 ( 0%) sys   0.00 ( 0%) wall       0 kB ( 0%) ggc
 tree aggressive DCE     :   0.01 ( 0%) usr   0.00 ( 0%) sys   0.02 ( 0%) wall       0 kB ( 0%) ggc
 tree rename SSA copies  :   0.01 ( 0%) usr   0.00 ( 0%) sys   0.01 ( 0%) wall       0 kB ( 0%) ggc
 dominance frontiers     :   0.06 ( 0%) usr   0.00 ( 0%) sys   0.04 ( 0%) wall       0 kB ( 0%) ggc
 dominance computation   :   0.80 ( 0%) usr   0.01 ( 0%) sys   0.88 ( 0%) wall       0 kB ( 0%) ggc
 out of ssa              :   0.49 ( 0%) usr   0.01 ( 0%) sys   0.65 ( 0%) wall    2496 kB ( 0%) ggc
 expand vars             :1260.37 (89%) usr   1.03 ( 7%) sys1272.57 (87%) wall   35735 kB ( 1%) ggc
 expand                  :   2.48 ( 0%) usr   0.66 ( 5%) sys   7.30 ( 0%) wall  589535 kB (22%) ggc
 post expand cleanups    :   0.35 ( 0%) usr   0.00 ( 0%) sys   0.46 ( 0%) wall      16 kB ( 0%) ggc
 varconst                :   0.01 ( 0%) usr   0.03 ( 0%) sys   0.04 ( 0%) wall     337 kB ( 0%) ggc
 forward prop            :   0.01 ( 0%) usr   0.00 ( 0%) sys   0.00 ( 0%) wall     161 kB ( 0%) ggc
 CSE                     :   0.01 ( 0%) usr   0.00 ( 0%) sys   0.01 ( 0%) wall      21 kB ( 0%) ggc
 dead code elimination   :   0.01 ( 0%) usr   0.00 ( 0%) sys   0.00 ( 0%) wall       0 kB ( 0%) ggc
 dead store elim1        :   0.01 ( 0%) usr   0.00 ( 0%) sys   0.01 ( 0%) wall     103 kB ( 0%) ggc
 dead store elim2        :   0.01 ( 0%) usr   0.00 ( 0%) sys   0.00 ( 0%) wall     126 kB ( 0%) ggc
 loop init               :   0.17 ( 0%) usr   0.00 ( 0%) sys   0.17 ( 0%) wall      13 kB ( 0%) ggc
 combiner                :   0.02 ( 0%) usr   0.00 ( 0%) sys   0.01 ( 0%) wall     231 kB ( 0%) ggc
 integrated RA           :  10.14 ( 1%) usr   0.50 ( 3%) sys  11.16 ( 1%) wall  290014 kB (11%) ggc
 LRA non-specific        :   3.70 ( 0%) usr   0.16 ( 1%) sys   4.33 ( 0%) wall    1116 kB ( 0%) ggc
 LRA virtuals elimination:   0.76 ( 0%) usr   0.02 ( 0%) sys   1.02 ( 0%) wall   40221 kB ( 2%) ggc
 LRA create live ranges  :   0.95 ( 0%) usr   0.03 ( 0%) sys   1.85 ( 0%) wall    5545 kB ( 0%) ggc
 LRA hard reg assignment :   0.07 ( 0%) usr   0.02 ( 0%) sys   0.09 ( 0%) wall       0 kB ( 0%) ggc
 reload                  :   0.08 ( 0%) usr   0.00 ( 0%) sys   0.09 ( 0%) wall       0 kB ( 0%) ggc
 reload CSE regs         :   0.01 ( 0%) usr   0.00 ( 0%) sys   0.01 ( 0%) wall     132 kB ( 0%) ggc
 thread pro- & epilogue  :   1.07 ( 0%) usr   0.00 ( 0%) sys   1.14 ( 0%) wall      71 kB ( 0%) ggc
 machine dep reorg       :   0.07 ( 0%) usr   0.01 ( 0%) sys   0.06 ( 0%) wall       0 kB ( 0%) ggc
 shorten branches        :   1.01 ( 0%) usr   0.00 ( 0%) sys   1.04 ( 0%) wall       0 kB ( 0%) ggc
 final                   :   2.61 ( 0%) usr   0.28 ( 2%) sys   6.93 ( 0%) wall   90145 kB ( 3%) ggc
 symout                  :   0.23 ( 0%) usr   0.06 ( 0%) sys   0.85 ( 0%) wall   51247 kB ( 2%) ggc
 uninit var analysis     :   0.06 ( 0%) usr   0.00 ( 0%) sys   0.06 ( 0%) wall       0 kB ( 0%) ggc
 rest of compilation     :   1.70 ( 0%) usr   0.07 ( 0%) sys   2.87 ( 0%) wall    4228 kB ( 0%) ggc
 unaccounted todo        :   0.00 ( 0%) usr   0.00 ( 0%) sys   0.04 ( 0%) wall       0 kB ( 0%) ggc
 repair loop structures  :   0.04 ( 0%) usr   0.00 ( 0%) sys   0.02 ( 0%) wall       0 kB ( 0%) ggc
 TOTAL                 :1421.11            14.31          1465.09            2637093 kB
```
Interesting to note with 6.4.0:
```
 phase opt and generate  :1410.95 (99%) usr   5.11 (36%) sys1445.52 (99%) wall 2286345 kB (87%) ggc
 expand vars             :1260.37 (89%) usr   1.03 ( 7%) sys1272.57 (87%) wall   35735 kB ( 1%) ggc
```
versus with 6.2.2:
```
expand vars             :   0.12 ( 0%) usr
```
Removing the `std.log` from the config clearly makes the compilation time back to normal.

Thanks

**Comment by nigoroll** [comment]
at least three :+1: for this excellent bug report from bugwash

**Comment by nigoroll** [comment]
FTR, we had #3230 which may or may not be related, but at any rate, it is pretty obvious that we are hitting bad scalability with the number of local variables.
Bugwash briefly discussed this and we want to look into two options:
* avoid local variables
* split functions generated from VCL subs into smaller chunks

I volunteered to look at this, and also the related question of a similar optimization for ACLs.

**Comment by n4n0r** [comment]
Great 👍 
If this can help, the compile time reduces to "normal" (just losing a few seconds) if the thousands of `std.log` are replaced by calls to a VCL subroutine containing this `std.log` only once, if it makes sense.

**Comment by nigoroll** [comment]
thank you for the update, @n4n0r - this is in line with the hypothesis
